src/Pos/Block/Core/Genesis/Misc.hs

```diff
@@ -96,3 +105,41 @@ instance HasDifficulty (GenesisBlock ssc) where
 
 instance BiHeader ssc => IsHeader (GenesisBlockHeader ssc)
 instance BiHeader ssc => IsGenesisHeader (GenesisBlockHeader ssc)
+
+----------------------------------------------------------------------------
+-- Smart constructors
+----------------------------------------------------------------------------
+
+type SanityConstraint ssc
+     = ( HasDifficulty $ BlockHeader ssc
+       , HasHeaderHash $ BlockHeader ssc
+       )
+
+-- | Smart constructor for 'GenesisBlockHeader'. Uses 'mkGenericHeader'.
+mkGenesisHeader
+    :: SanityConstraint ssc
+    => Maybe (BlockHeader ssc)
+    -> EpochIndex
+    -> Body (GenesisBlockchain ssc)
+    -> GenesisBlockHeader ssc
+mkGenesisHeader prevHeader epoch body =
+    mkGenericHeader prevHeader body consensus (GenesisExtraHeaderData $ mkAttributes ())
+  where
+    difficulty = maybe 0 (view difficultyL) prevHeader
+    consensus _ _ =
+        GenesisConsensusData {_gcdEpoch = epoch, _gcdDifficulty = difficulty}
+
+-- | Smart constructor for 'GenesisBlock'.
+mkGenesisBlock
+    :: SanityConstraint ssc
+    => Maybe (BlockHeader ssc)
+    -> EpochIndex
+    -> SlotLeaders
+    -> GenesisBlock ssc
+mkGenesisBlock prevHeader epoch leaders =
+    either disaster identity $ recreateGenericBlock header body extra
+  where
+    header = mkGenesisHeader prevHeader epoch body
+    body = GenesisBody leaders
+    extra = GenesisExtraBodyData $ mkAttributes ()
+    disaster = error . mappend "mkGenesisBlock: "
```

Comment:
I guess it's better to use explicit type-level exceptions here (TODO next part of this refactoring).